### PR TITLE
intelligent scissor JS sample 

### PR DESCRIPTION
**Pull request by SimpleVlad**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under OpenCV (BSD) License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders_only=Docs
build_image:Docs=docs-js
```

### CODE REVIEW

File: doc/js_tutorials/intelligent_scissor.html [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+</canvas>
+<canvas id="5" disabled=true>
+</canvas>
+<img id="src" src="fruits.jpg"/>
+<input type="button" id="myButton" value="Test" onclick="" />
+<input type="button" id="startAlgo" value="Start" disabled=true onclick="main()">
+<canvas id="canvas2" disabled=true>
+</canvas>
+<canvas id="canvas3" disabled=true>
+</canvas>
```
Comment (dkurt) [inline doc/js_tutorials/intelligent_scissor.html:26]: please format with alignment

File: doc/js_tutorials/intelligent_scissor.html [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+  }
+
+  return WEIGHT_LAP_ZERO_CROSS * zero_crossing.floatAt(q.x, q.y) +  WEIGHT_GRADIENT_DIRECTION * (Math.cos(dp) + Math.cos(dq) / Math.PI) + WEIGHT_GRADIENT_MAGNITUDE *fG;
+}
+
+function find_min_path(start)
+{
+  let begin;
+  
+  let cost_map = new cv.Mat(src.cols, src.rows, cv.CV_32F, new cv.Scalar(10^10));
```
Comment (dkurt) [inline doc/js_tutorials/intelligent_scissor.html:126]: What is 10^10?

File: doc/js_tutorials/intelligent_scissor.html [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    enqueue(obj, level)
+    {
+        let i = 0;
+        while(i < this.queue.length && this.queue[i].level < level)
+            ++i;
+        this.queue.splice(i,0,{
+            value : obj,
+            level
+        });
+    }
```
Comment (dkurt) [inline doc/js_tutorials/intelligent_scissor.html:57]: Please keep 2 whitespace everywhere and opening `{` on the same line:
```js
enqueue(obj, level) {
  let i = 0;
```

File: doc/js_tutorials/intelligent_scissor.html [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    dq = Iy.floatAt(q.x, q.y) * (p.x - q.x) - Ix.floatAt(q.x, q.y) * (p.y - q.y);
+  }
+  if (isDiag)
+  {
+    dp /= Math.sqrt(2);
+    dq /= Math.sqrt(2);
+  }
+  else
+  {
+    fG /= Math.sqrt(2);
```
Comment (dkurt) [inline doc/js_tutorials/intelligent_scissor.html:116]: Please keep it as consant

File: doc/js_tutorials/intelligent_scissor.html [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    }
+    cv.imshow("canvas",  src);
+  }
+}
+
+function main()
+{
+  src = cv.imread("src");
+  cv.imshow("canvas", src);
+  cv.imshow("canvas", src);
```
Comment (dkurt) [inline doc/js_tutorials/intelligent_scissor.html:191]: double `imshow`

File: doc/js_tutorials/intelligent_scissor.html [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+  zero_crossing = new cv.Mat();
+  img_canny = new cv.Mat();
+  grayscale = new cv.Mat();
+  Ix = new cv.Mat();
+  Iy = new cv.Mat();
+  
+  gradient_magnitude = new cv.Mat();
+  cv.cvtColor(src, grayscale, cv.COLOR_BGR2GRAY);
```
Comment (dkurt) [inline doc/js_tutorials/intelligent_scissor.html:199]: origin image is RGBA

**Comment by dkurt** [comment]
Kind reminder

File: doc/js_tutorials/intelligent_scissor.html (hunk 6 above)
Comment (dkurt) [reply doc/js_tutorials/intelligent_scissor.html]: @SimpleVlad

File: doc/js_tutorials/intelligent_scissor.html [hunk 7]
```diff
@@ -78,20 +78,20 @@
...
-          dq = Iy.floatAt(q.x, q.y) * (q.x - p.x) - Ix.floatAt(q.x, q.y) * (q.y - p.y);
+          dp = Iy.data32F[p.x * Iy.cols * Iy.channels() + p.y * Iy.channels()] * (q.x - p.x) - Ix.data32F[p.x * Ix.cols * Ix.channels() + p.y * Ix.channels()] * (q.y - p.y);
+          dq = Iy.data32F[q.x * Iy.cols * Iy.channels() + q.y * Iy.channels()] * (q.x - p.x) - Ix.data32F[q.x * Ix.cols * Ix.channels() + q.y * Ix.channels()] * (q.y - p.y);
         }
         else
         {
-          dp = Iy.floatAt(p.x, p.y) * (p.x - q.x) - Ix.floatAt(p.x, p.y) * (p.y - q.y);
-          dq = Iy.floatAt(q.x, q.y) * (p.x - q.x) - Ix.floatAt(q.x, q.y) * (p.y - q.y);
+          dp = Iy.data32F[p.x * Iy.cols * Iy.channels() + p.y * Iy.channels()] * (p.x - q.x) - Ix.data32F[p.x * Ix.cols * Ix.channels() + p.y * Ix.channels()] * (p.y - q.y);
+          dq = Iy.data32F[q.x * Iy.cols * Iy.channels() + q.y * Iy.channels()] * (p.x - q.x) - Ix.data32F[q.x * Ix.cols * Ix.channels() + q.y * Ix.channels()] * (p.y - q.y);
```
Comment (dkurt) [inline doc/js_tutorials/intelligent_scissor.html:94]: Please save it to some local variable:
```
let Iy_data = Iy.data32F;
let Ix_data = Ix.data32F;
```

File: doc/js_tutorials/intelligent_scissor.html [hunk 8]
```diff
@@ -180,7 +179,7 @@
         Iy = new cv.Mat();
         
         gradient_magnitude = new cv.Mat();
-        cv.cvtColor(src, grayscale, cv.COLOR_BGR2GRAY);
+        cv.cvtColor(src, grayscale, cv.COLOR_BGRA2GRAY);
```
Comment (dkurt) [inline doc/js_tutorials/intelligent_scissor.html:182]: Image is RGBA, not BGRA

**Comment by SimpleVlad** [comment]
**cpp sample with BGR2BGRA:**
`[20, 20] 0
loc 1
loc 0.958995
loc 1
loc 0.958995
loc 0.958995
loc 1
loc 0.958995
loc 1
[19, 20] 0.958995
loc 1.95899
loc 1.91799
loc 1.95899
loc 1.91799
loc 1.91799
loc 1.95899
loc 1.95899
[21, 20] 0.958995
loc 1.95899
loc 1.95899
loc 1.91799
loc 1.91799
loc 1.95807
loc 1.91761
loc 1.95883
[20, 19] 0.958995
loc 1.95899
loc 1.91799
loc 1.91799
loc 1.95899
loc 1.91799
[20, 21] 0.958995
loc 1.91799
loc 1.95899
loc 1.91799
loc 1.91799
loc 1.95899
[19, 21] 1
loc 2
loc 1.95899
loc 2
loc 1.95899
loc 2
[21, 21] 1
loc 2
loc 1.95899
loc 1.99908
loc 1.95861
loc 1.99984
[21, 19] 1
loc 2
loc 1.95899
loc 1.99908
loc 1.95861
loc 1.99984
[19, 19] 1
loc 2
loc 1.95899
loc 2
loc 1.95899
loc 2
[22, 20] 1.91761`

**original cpp sample:**
`[20, 20] 0
loc 1
loc 0.958995
loc 1
loc 0.958995
loc 0.958995
loc 1
loc 0.958995
loc 1
[19, 20] 0.958995
loc 1.95899
loc 1.91799
loc 1.95899
loc 1.91799
loc 1.91799
loc 1.95899
loc 1.95899
[21, 20] 0.958995
loc 1.95899
loc 1.95899
loc 1.91799
loc 1.91799
loc 1.95807
loc 1.91761
loc 1.95883
[20, 19] 0.958995
loc 1.95899
loc 1.91799
loc 1.91799
loc 1.95899
loc 1.91799
[20, 21] 0.958995
loc 1.91799
loc 1.95899
loc 1.91799
loc 1.91799
loc 1.95899
[19, 21] 1
loc 2
loc 1.95899
loc 2
loc 1.95899
loc 2
[21, 21] 1
loc 2
loc 1.95899
loc 1.99908
loc 1.95861
loc 1.99984
[21, 19] 1
loc 2
loc 1.95899
loc 1.99908
loc 1.95861
loc 1.99984
[19, 19] 1
loc 2
loc 1.95899
loc 2
loc 1.95899
loc 2
[22, 20] 1.91761`

**js:**
```
x: 21, y: 20}
smpl.html:142 0.9589939713478088
2 smpl.html:163 1.9589939713478088
2 smpl.html:163 1.917987971347809
smpl.html:163 1.9589939713478088
smpl.html:163 1.917987971347809
smpl.html:163 1.9589939713478088
smpl.html:140 {x: 20, y: 21}
smpl.html:142 0.9589939713478088
smpl.html:163 1.9589939713478088
smpl.html:163 1.917987971347809
smpl.html:163 1.9588329713478088
smpl.html:163 1.9181429713478089
smpl.html:163 1.917987971347809
smpl.html:163 1.9588329713478088
smpl.html:140 {x: 20, y: 19}
smpl.html:142 0.9589939713478088
smpl.html:163 1.9589939713478088
smpl.html:163 1.917987971347809
smpl.html:163 1.9589939713478088
smpl.html:163 1.917987971347809
smpl.html:163 1.9589939713478088
smpl.html:163 1.917987971347809
smpl.html:140 {x: 19, y: 20}
smpl.html:142 0.9589939713478088
smpl.html:163 1.9589939713478088
smpl.html:163 1.917987971347809
smpl.html:163 1.9589939713478088
2 smpl.html:163 1.917987971347809
smpl.html:140 {x: 21, y: 21}
smpl.html:142 1
smpl.html:163 1.999839
smpl.html:163 1.959149
smpl.html:163 2
smpl.html:163 1.9589940000000001
smpl.html:163 1.999839
smpl.html:140 {x: 21, y: 19}
smpl.html:142 1
smpl.html:163 2
smpl.html:163 1.9589940000000001
smpl.html:163 2
smpl.html:163 1.9589940000000001
smpl.html:163 2
smpl.html:140 {x: 19, y: 21}
smpl.html:142 1
smpl.html:163 2
smpl.html:163 1.9589940000000001
smpl.html:163 1.999839
smpl.html:163 1.959149
smpl.html:163 1.999839
smpl.html:140 {x: 19, y: 19}
smpl.html:142 1
```

**Comment by SimpleVlad** [comment]
priority_queue js  in a single neiborhood
```
smpl.html:187 {x: 19, y: 20} 0.9589949250221252
smpl.html:187 {x: 20, y: 19} 0.9589949250221252
smpl.html:187 {x: 20, y: 21} 0.9589949250221252
smpl.html:187 {x: 21, y: 20} 0.9589949250221252
smpl.html:187 {x: 19, y: 19} 1
smpl.html:187 {x: 19, y: 21} 1
smpl.html:187 {x: 21, y: 19} 1
smpl.html:187 {x: 21, y: 21} 1
```
priority_queue c++  in a single neiborhood
```
[19, 20] 0.958995
[21, 20] 0.958995
[20, 19] 0.958995
[20, 21] 0.958995
[19, 21] 1
[21, 19] 1
[21, 21] 1
[19, 19] 1
```

**Comment by SimpleVlad** [comment]
```

smpl.html:221 Point {x: 129, y: 34}
smpl.html:221 Point {x: 125, y: 35}
smpl.html:227 Point {x: 36, y: 124}
smpl.html:227 Point {x: 125, y: 35}
smpl.html:227 Point {x: 36, y: 124}
smpl.html:227 Point {x: 125, y: 35}
smpl.html:227 Point {x: 36, y: 124}
smpl.html:227 Point {x: 125, y: 35}
smpl.html:227 Point {x: 36, y: 124}
smpl.html:227 Point {x: 125, y: 35}
smpl.html:227 Point {x: 36, y: 124}
```

**Comment by SimpleVlad** [comment]
```
python ./platforms/js/build_js.py build_js --emscripten_dir ~/emsdk/fastcomp/emscripten/ 
Args: Namespace(build_dir='build_js', build_doc=False, build_flags=None, build_perf=False, build_test=False, build_wasm=False, build_wasm_intrin_test=False, clean_build_dir=False, cmake_option=None, config='/home/vlad/opencv2/opencv/platforms/js/opencv_js.config.py', config_only=False, disable_wasm=False, emscripten_dir='/home/vlad/emsdk/fastcomp/emscripten/', enable_exception=False, opencv_dir='/home/vlad/opencv2/opencv', simd=False, skip_config=False, threads=False)
Check dir /home/vlad/opencv2/opencv/build_js (create: True, clean: False)
Check dir /home/vlad/opencv2/opencv (create: False, clean: False)
Check dir /home/vlad/emsdk/fastcomp/emscripten (create: False, clean: False)
=====
===== Config OpenCV.js build for default target
=====
Executing: ['cmake', '-DENABLE_PIC=FALSE', '-DCMAKE_BUILD_TYPE=Release', "-DCMAKE_TOOLCHAIN_FILE='/home/vlad/emsdk/fastcomp/emscripten/cmake/Modules/Platform/Emscripten.cmake'", "-DCPU_BASELINE=''", "-DCPU_DISPATCH=''", '-DCV_TRACE=OFF', '-DBUILD_SHARED_LIBS=OFF', '-DWITH_1394=OFF', '-DWITH_VTK=OFF', '-DWITH_EIGEN=OFF', '-DWITH_FFMPEG=OFF', '-DWITH_GSTREAMER=OFF', '-DWITH_GTK=OFF', '-DWITH_GTK_2_X=OFF', '-DWITH_IPP=OFF', '-DWITH_JASPER=OFF', '-DWITH_JPEG=OFF', '-DWITH_WEBP=OFF', '-DWITH_OPENEXR=OFF', '-DWITH_OPENGL=OFF', '-DWITH_OPENVX=OFF', '-DWITH_OPENNI=OFF', '-DWITH_OPENNI2=OFF', '-DWITH_PNG=OFF', '-DWITH_TBB=OFF', '-DWITH_TIFF=OFF', '-DWITH_V4L=OFF', '-DWITH_OPENCL=OFF', '-DWITH_OPENCL_SVM=OFF', '-DWITH_OPENCLAMDFFT=OFF', '-DWITH_OPENCLAMDBLAS=OFF', '-DWITH_GPHOTO2=OFF', '-DWITH_LAPACK=OFF', '-DWITH_ITT=OFF', '-DWITH_QUIRC=OFF', '-DBUILD_ZLIB=ON', '-DBUILD_opencv_apps=OFF', '-DBUILD_opencv_calib3d=ON', '-DBUILD_opencv_dnn=ON', '-DBUILD_opencv_features2d=ON', '-DBUILD_opencv_flann=OFF', '-DBUILD_opencv_ml=OFF', '-DBUILD_opencv_photo=OFF', '-DBUILD_opencv_imgcodecs=OFF', '-DBUILD_opencv_shape=OFF', '-DBUILD_opencv_videoio=OFF', '-DBUILD_opencv_videostab=OFF', '-DBUILD_opencv_highgui=OFF', '-DBUILD_opencv_superres=OFF', '-DBUILD_opencv_stitching=OFF', '-DBUILD_opencv_java=OFF', '-DBUILD_opencv_java_bindings_generator=OFF', '-DBUILD_opencv_js=ON', '-DBUILD_opencv_python2=OFF', '-DBUILD_opencv_python3=OFF', '-DBUILD_opencv_python_bindings_generator=OFF', '-DBUILD_EXAMPLES=OFF', '-DBUILD_PACKAGE=OFF', '-DBUILD_TESTS=OFF', '-DBUILD_PERF_TESTS=OFF', '-DBUILD_DOCS=OFF', '-DWITH_PTHREADS_PF=OFF', '-DCV_ENABLE_INTRINSICS=OFF', '-DBUILD_WASM_INTRIN_TESTS=OFF', "-DCMAKE_C_FLAGS='-s USE_PTHREADS=0 '", "-DCMAKE_CXX_FLAGS='-s USE_PTHREADS=0 '", '/home/vlad/opencv2/opencv']
Re-run cmake no build system arguments
CMake Error at /usr/local/share/cmake-3.15/Modules/CMakeDetermineSystem.cmake:99 (message):
  Could not find toolchain file:
  /home/vlad/emsdk/fastcomp/emscripten/cmake/Modules/Platform/Emscripten.cmake
Call Stack (most recent call first):
  CMakeLists.txt:106 (enable_language)


CMake Error: CMake was unable to find a build program corresponding to "Unix Makefiles".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
Traceback (most recent call last):
  File "./platforms/js/build_js.py", line 267, in <module>
    builder.config()
  File "./platforms/js/build_js.py", line 190, in config
    execute(cmd)
  File "./platforms/js/build_js.py", line 23, in execute
    raise Fail("Child returned: %s" % retcode)
__main__.Fail: Child returned: 1`
```

**Comment by asmorkalov** [comment]
@SimpleVlad What is the patch status? Do you plan to finish it im mean time. CI bot reports issue with trailing whitespaces and JS build errors.

**Comment by alalek** [comment]
JavaScript sample is replaced by #19194
It would be nice to update C++ sample to use API from OpenCV.

**Pull request closed without merging**